src/ewah.zig

```diff
@@ -0,0 +1,252 @@
+        assert(std.math.maxInt(MarkerUniformCount) == marker_uniform_word_count_max);
+
+        assert(@bitSizeOf(MarkerLiteralCount) % 2 == 0);
+        assert(std.math.maxInt(MarkerLiteralCount) == marker_literal_word_count_max);
+    }
+
+    return struct {
+        const Self = @This();
+
+        inline fn marker_word(mark: Marker) Word {
+            return @bitCast(Word, mark);
+        }
+
+        /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+        /// Returns the number of *words* written to `target_words`.
+        pub fn decode(source: []const u8, target_words: []Word) usize {
+            assert(source.len % @sizeOf(Word) == 0);
+            const source_words = @alignCast(@alignOf(Word), mem.bytesAsSlice(Word, source));
+            var source_index: usize = 0;
+            var target_index: usize = 0;
+            while (source_index < source_words.len) {
+                const marker = @bitCast(Marker, source_words[source_index]);
+                const uniform_word_count: usize = marker.uniform_word_count;
+                const uniform_word: Word = if (marker.uniform_bit == 1) ~@as(Word, 0) else 0;
+                const literal_word_count: usize = marker.literal_word_count;
+                std.mem.set(Word, target_words[target_index..][0..uniform_word_count], uniform_word);
+                std.mem.copy(Word, target_words[target_index + uniform_word_count..][0..literal_word_count],
+                    source_words[source_index + 1..][0..literal_word_count]);
+                target_index += uniform_word_count + literal_word_count;
+                source_index += 1 + literal_word_count;
+            }
+            assert(source_index == source_words.len);
+            assert(target_index <= target_words.len);
+            return target_index;
+        }
+
+        // Returns the number of bytes written to `target`.
+        pub fn encode(source_words: []const Word, target: []u8) usize {
+            assert(target.len >= @sizeOf(Marker));
+            assert(target.len == encode_size_max(source_words));
+
+            const target_words = @alignCast(@alignOf(Word), mem.bytesAsSlice(Word, target));
+            std.mem.set(Word, target_words, 0);
+
+            var target_index: usize = 0;
+            var source_index: usize = 0;
+            while (source_index < source_words.len) {
+                const word = source_words[source_index];
+                const uniform_word_count = count: {
+                    if (is_literal(word)) break :count 0;
+                    // Measure run length.
+                    const uniform_max = std.math.min(source_words.len, source_index + marker_uniform_word_count_max);
+                    var i: usize = source_index + 1;
+                    while (i < uniform_max and source_words[i] == word) : (i += 1) {}
+                    const uniform = i - source_index;
+                    break :count uniform;
```

Comment:
It was bugging me that `uniform_max` wasn't the maximum uniform word count we are calculating but rather the maximum uniform word count plus the current source index. I tried fixing this and I think it makes the code a bit cleaner, let me know what you think:
```suggestion
                    const uniform_max = math.min(source_words.len - source_index, marker_uniform_word_count_max);
                    var uniform: usize = 1;
                    while (uniform < uniform_max and source_words[source_index + uniform] == word) uniform += 1;
                    break :count uniform;
```